lib/core/include/scipp/core/dict.h

```diff
@@ -0,0 +1,423 @@
+      : m_iterators{std::forward<T>(it1)}, m_container(container),
+        m_end_address(container.get().data() + container.get().size()) {}
+
+  template <class T, class U>
+  Iterator(std::reference_wrapper<Container> container, T &&it1, U &&it2)
+      : m_iterators{std::forward<T>(it1), std::forward<U>(it2)},
+        m_container(container),
+        m_end_address(container.get().data() + container.get().size()) {}
+
+  decltype(auto) operator*() const {
+    expect_container_unchanged();
+    if constexpr (sizeof...(IteratorIndices) == 1) {
+      return *std::get<0>(m_iterators);
+    } else {
+      return std::make_pair(std::cref(*std::get<0>(m_iterators)),
+                            std::ref(*std::get<1>(m_iterators)));
+    }
+  }
+
+  decltype(auto) operator->() const {
+    if constexpr (sizeof...(IteratorIndices) == 1) {
+      expect_container_unchanged();
+      return std::get<0>(m_iterators);
+    } else {
+      return TemporaryItem<reference>(**this);
+    }
+  }
+
+  Iterator &operator++() {
+    expect_container_unchanged();
+    (++std::get<IteratorIndices>(m_iterators), ...);
+    return *this;
+  }
+
+  bool operator==(
+      const Iterator<Container, It1, It2, IteratorIndices...> &other) const {
+    expect_container_unchanged();
+    // Assuming m_iterators are always in sync.
+    return std::get<0>(m_iterators) == std::get<0>(other.m_iterators);
+  }
+
+  bool operator!=(
+      const Iterator<Container, It1, It2, IteratorIndices...> &other) const {
+    return !(*this == other);
+  }
+
+  template <class F> auto transform(F &&func) const & {
+    return TransformIterator{*this, std::forward<F>(func)};
+  }
+
+  template <class F> auto transform(F &&func) && {
+    return TransformIterator{std::move(*this), std::forward<F>(func)};
+  }
+
+  friend void swap(Iterator &a, Iterator &b) {
+    swap(a.m_iterators, b.m_iterators);
+    swap(a.m_container, b.m_container);
+    std::swap(a.m_end_address, b.m_end_address);
+  }
+
+protected:
+  // operator-> needs to return a pointer or something that has operator->
+  // But we cannot take the address of the temporary pair or transform result.
+  // So store it in this wrapper to make it accessible via its address.
+  template <class T> class TemporaryItem {
+  public:
+    explicit TemporaryItem(T &&item) : m_item(std::move(item)) {}
+    auto *operator->() { return &m_item; }
+
+  private:
+    std::decay_t<T> m_item;
+  };
+
+private:
+  using IteratorStorage =
+      std::tuple<typename std::tuple_element<IteratorIndices,
+                                             std::tuple<It1, It2>>::type...>;
+
+  IteratorStorage m_iterators;
+  std::reference_wrapper<Container> m_container;
+  const void *m_end_address;
+
+  void expect_container_unchanged() const {
+    if (m_container.get().data() + m_container.get().size() != m_end_address) {
+      throw std::runtime_error("dictionary changed size during iteration");
```

Comment:
How would this cause a problem? Suppose we have a length=1 vector when we create the iterator and suppose `data()` is 0x10. Then, `m_end_address = data() + size() = 0x11`. When the element is erased, there are multiple possible outcomes:
- `data` returns null -> `data() + size() = 0x00 + 0 = 0x00 != m_end_address`
- `data` returns the previous base address -> `data() + size() = 0x10 + 0 = 0x10 != m_end_address`
- `data` returns the previous end address -> `data() + size() = 0x11 + 0 = 0x11 = m_end_address` This case would cause a problem. But is there any real risk that this will happen?

Or the other way around. Suppose we start with an empty vector and add an element:
- `data` of empty returns 0x00 -> after push: `data() + size() = 0x10 + 1 = 0x11 != m_end_address`
- `data` of empty returns 0x11 -> after push: `data() + size() = 0x10 + 1 = 0x11 = m_end_address` This would require the pointer of empty to start one byte after the location where memory will be allocated. This will never happen in practice unless std::vector can see the future.

If there is a special allocator, e.g. using a memory pool, then there could be problems because a vector might reallocate and end up in the same spot. But do we need to worry about this?

Reply:
> But is there any real risk that this will happen?

I don't know, is there a reason to take any chances? The extra pointer comparison would not add any real overhead, would it?

> If there is a special allocator, e.g. using a memory pool, then there could be problems because a vector might reallocate and end up in the same spot. But do we need to worry about this?

Yes, absolutely! The allocators used by the standard libraries have become quite advanced. For example, as far as I know gcc's allocator now includes many things that `tcmalloc` used to do, by default.